dipy/nn/utils.py

```diff
@@ -92,84 +100,196 @@ def transform_img(image, affine, *, voxsize=None, init_shape=(256, 256, 256), sc
+    image2 : np.ndarray
+        Recovered image
     """
+    new_affine = np.linalg.inv(inv_affine)
     new_image, _ = reslice(image, np.eye(4), (scale, scale, scale), (1, 1, 1))
-    recovered_img = affine_transform(new_image, affine, output_shape=ori_shape)
-    affine[:3, 3] += np.array(
-        [init_shape[0] // 2, init_shape[1] // 2, init_shape[2] // 2]
+    crop_vs = crop_vs.astype(int)
+    pad_vs = pad_vs.astype(int)
+    new_image = np.pad(
+        new_image,
+        (
+            (crop_vs[0, 0], crop_vs[0, 1]),
+            (crop_vs[1, 0], crop_vs[1, 1]),
+            (crop_vs[2, 0], crop_vs[2, 1]),
+        ),
     )
+    new_image = new_image[
+        pad_vs[0, 0] : new_image.shape[0] - pad_vs[0, 1],
+        pad_vs[1, 0] : new_image.shape[1] - pad_vs[1, 1],
+        pad_vs[2, 0] : new_image.shape[2] - pad_vs[2, 1],
+    ]
+    new_image = affine_transform(new_image, new_affine, output_shape=mid_shape)
+    affine = np.matmul(np.linalg.inv(offset_array), new_affine)
     if voxsize is not None and np.any(voxsize != np.ones(3)):
-        kwargs = {
-            "order": 1,
-            "mode": "constant",
-            "matrix": voxsize,
-            "output_shape": image_shape,
-        }
-        recovered_img = np.round(affine_transform(recovered_img, **kwargs))
-    return recovered_img
+        image2, _ = reslice(new_image, affine, (1, 1, 1), voxsize)
+    else:
+        image2 = new_image
+
+    # because of zoom rounding errors
+    image2, _, _ = pad_crop(image2, ori_shape)
+    return image2
+
+
+def pad_crop(image, target_shape):
+    """
+    Function to figure out pad and crop range
+    to fit the target shape with the image
+
+    Parameters
+    ----------
+    image : np.ndarray
+        Target image
+    target_shape : (3,)
+        Target shape
+
+    Returns
+    -------
+    image : np.ndarray
+        Padded/cropped image
+    pad_vs : np.ndarray (3,2)
+        Pad range used
+    crop_vs : np.ndarray (3,2)
+        Crop range used
+    """
+    crop_vs = np.zeros((3, 2)).astype(int)
+    pad_vs = np.zeros((3, 2)).astype(int)
+    if image.shape != target_shape:
+        pad_crop_v = np.array(target_shape) - np.array(image.shape)
+        for d in range(3):
+            if pad_crop_v[d] < 0:
+                crop_vs[d, 0] = -pad_crop_v[d] // 2
+                crop_vs[d, 1] = -pad_crop_v[d] - crop_vs[d, 0]
+            elif pad_crop_v[d] > 0:
+                pad_vs[d, 0] = pad_crop_v[d] // 2
+                pad_vs[d, 1] = pad_crop_v[d] - pad_vs[d, 0]
+        image = np.pad(
+            image,
+            (
+                (pad_vs[0, 0], pad_vs[0, 1]),
+                (pad_vs[1, 0], pad_vs[1, 1]),
+                (pad_vs[2, 0], pad_vs[2, 1]),
+            ),
+        )
+        image = image[
+            crop_vs[0, 0] : image.shape[0] - crop_vs[0, 1],
+            crop_vs[1, 0] : image.shape[1] - crop_vs[1, 1],
+            crop_vs[2, 0] : image.shape[2] - crop_vs[2, 1],
+        ]
+
```

Comment:
this function is mainly indented so I will think the other way to avoid all this indentation meaning

```python
if image.shape == target_shape:
    return image, pad_vs, crop_vs

pad_crop_v = np.array(target_shape) - np.array(image.shape)
....
```